### PR TITLE
KAFKA-6718 / Add rack awareness configurations to StreamsConfig

### PR DESCRIPTION
**Pull request by lkokhreidze**
This PR is part of [KIP-708](https://cwiki.apache.org/confluence/display/KAFKA/KIP-708%3A+Rack+awareness+for+Kafka+Streams) and adds rack aware standby task assignment logic.

Rack aware standby task assignment won't be functional until all parts of this KIP gets merged.

Splitting PRs into three smaller PRs to make the review process easier to follow. Overall plan is the following:

⏭️  Rack aware standby task assignment logic https://github.com/apache/kafka/pull/10851
⏭️  Protocol change, add clientTags to SubscriptionInfoData https://github.com/apache/kafka/pull/10802
👉  Add required configurations to StreamsConfig (public API change, at this point we should have full functionality)

This PR implements last point of the above mentioned plan.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by cadonna** [comment]
@lkokhreidze Is this PR ready for review?

**Comment by lkokhreidze** [comment]
Hi @cadonna 
Not yet. Will open it for review by Monday.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (lkokhreidze) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: Do not have any science behind those numbers, but thought it's better to have some restrictions.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (lkokhreidze) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:1197]: Not sure if this is the best place to perform those validations, but couldn't find any better place.

File: streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java
Comment (lkokhreidze) [inline streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java:63]: Since core logic tests are already written as unit tests, this integration tests verifies basic high level functionality.

**Comment by lkokhreidze** [comment]
Ready for the review @cadonna @showuon

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/HighAvailabilityTaskAssignor.java
Comment (lkokhreidze) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/HighAvailabilityTaskAssignor.java:146]: Refactored based on the previous PR comment https://github.com/apache/kafka/pull/10851#discussion_r817650626

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StandbyTaskAssignmentUtils.java
Comment (lkokhreidze) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StandbyTaskAssignmentUtils.java]: Refactored based on previous PR comment
https://github.com/apache/kafka/pull/10851#discussion_r817629787

Since both standby task assignors have the same logging, moved the logging to this utility method.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientTagAwareStandbyTaskAssignor.java
Comment (lkokhreidze) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientTagAwareStandbyTaskAssignor.java:213]: Added logging based on the previous PR comment https://github.com/apache/kafka/pull/10851#discussion_r817629787

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientTagAwareStandbyTaskAssignor.java
Comment (lkokhreidze) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientTagAwareStandbyTaskAssignor.java:201]: This was not requested during the previous PR review, but I think it can be useful for debugging to have extra logs.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java
Comment (cadonna) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Could you rename this method to somthing a bit more specific like `atMostOfLength()`?

File: clients/src/test/java/org/apache/kafka/common/config/ConfigDefTest.java
Comment (cadonna) [inline clients/src/test/java/org/apache/kafka/common/config/ConfigDefTest.java:760]: Could you add a unit test, so that there is one with a list that is longer than the maximum, one with a list that is strictly shorter than maximum, and one with list that has maximum length?

**Comment by lkokhreidze** [comment]
There are few tests failing. Will be fixing them now.

File: streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java:209]: I think, here it makes sense to wait until all Streams clients are `RUNNING` so that we know that the rebalance is done.

File: streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java:154]: See my comment below about waiting for `RUNNING`.

File: streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java:218]: Could you add a test where standbys cannot be distributed over all tags to check that task assignment works in that case?

**Review by cadonna — COMMENTED** [review]
@lkokhreidze Thanks for the PR!

Here my comments.

**Comment by lkokhreidze** [comment]
Thanks @cadonna for the feedback.
I've addressed your comments.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:1468]: This seems to be only used in its own test.

Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:1468]: Sorry, my bad. I misinterpreted the code.

Comment (lkokhreidze) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:1468]: It is indeed only used in tests. But seems like that's a common pattern in the `StreamsPartitionAssignor`. Followed the same approach. Happy to remove this if you think it's unnecessary.

**Comment by lkokhreidze** [comment]
Ran `./gradlew :streams:testAll` locally. The only test that failed was `EosIntegrationTest. shouldCommitCorrectOffsetIfInputTopicIsTransactional[at_least_once]` but seems like a flaky test.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (showuon) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: Actually, I don't think we need this limitation. Could we not be able to support list size > 10, key len > 50, value len > 100? I guess we can. If so, why this limitation? And like what you've asked, why this number?

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (showuon) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:1197]: Yes, put validation here is good.

Comment (showuon) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:1197]: Maybe put the rack awareness validation into an another method. (If we decided to have these limitation)

File: streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java
Comment (showuon) [inline streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java]: We already have `startApplicationAndWaitUntilRunning` or `waitForApplicationState` in `IntegrationTestUtils`. I think we can use `startApplicationAndWaitUntilRunning` in our tests.

File: streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java
Comment (showuon) [inline streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java]: Since we've confirmed the stream clients are in running state, should we also `wait` for the tasks distribution? I think we can directly verify them, right?

**Review by showuon — COMMENTED** [review]
@lkokhreidze , thanks for the PR. Left some comments.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (lkokhreidze) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: > Actually, I don't think we need this limitation. Could we not be able to support list size > 10, key len > 50, value len > 100?

I think we need some limitations because this information will be encoded into the subscription info and subscription info is the part of the group metadata message that we need to write into the internal topic. This was also discussed in the voting thread.

As for why those numbers, I really can't say. It can be more, can be less. But we need some limitation I think. But I can't imagine valid use case of having more than 10 tags with more than 50 and 100 characters length.

File: streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java
Comment (lkokhreidze) [reply streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java]: Yes good point. Will update the test.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (showuon) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: Make sense to me. Could we add a comment above these line, to mention we have these limitation because these tag value will be encoded into subscription info, sth like that? Thanks.

**Comment by lkokhreidze** [comment]
Thanks for the review @showuon 
I've addressed your comments.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StandbyTaskAssignmentUtils.java
Comment (showuon) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StandbyTaskAssignmentUtils.java]: nit: `{}`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StandbyTaskAssignmentUtils.java
Comment (showuon) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StandbyTaskAssignmentUtils.java]: This doesn't look right. We log the warning to user even when `numRemainingStandbys==0`?

**Review by showuon — COMMENTED** [review]
Thanks for the update. Left some more comments.

File: clients/src/test/java/org/apache/kafka/common/config/ConfigDefTest.java
Comment (showuon) [inline clients/src/test/java/org/apache/kafka/common/config/ConfigDefTest.java]: We can use `assertDoesNotThrow` here and below.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java:1140]: I'm not quite sure if the `value` will output the list strings. Should we use `values` instead?

**Review by showuon — COMMENTED** [review]
Made another pass. Left a comment. Thanks.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java
Comment (lkokhreidze) [reply clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java:1140]: It should output list of strings. Added test to verify.

File: clients/src/test/java/org/apache/kafka/common/config/ConfigDefTest.java
Comment (lkokhreidze) [reply clients/src/test/java/org/apache/kafka/common/config/ConfigDefTest.java]: Done

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StandbyTaskAssignmentUtils.java
Comment (lkokhreidze) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StandbyTaskAssignmentUtils.java]: Stupid mistake. Fixed. Also added tests.

**Comment by lkokhreidze** [comment]
Hi @showuon 
Thanks for taking a look. I've addressed your comments.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: I am a bit concerned about the length. I would propose the following:

```
MAX_RACK_AWARE_ASSIGNMENT_TAG_LIST_SIZE = 5
MAX_RACK_AWARE_ASSIGNMENT_TAG_KEY_LENGTH = 20
MAX_RACK_AWARE_ASSIGNMENT_TAG_VALUE_LENGTH = 30
```

We can always increase these numbers in future if they are too small, but if we want to decrease them, we might run into backwards compatibility issues.
 
But more important, we should have an integration test that tests the maximum allowed size.

**Review by cadonna — COMMENTED** [review]
@lkokhreidze Thank you for the updates!

I have one major comment.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (lkokhreidze) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: Thanks for the comment. I will add the integration test to test all three of those configurations.

**Comment by lkokhreidze** [comment]
Thanks @cadonna for the review.
I've addressed your comment in this commit https://github.com/apache/kafka/pull/11837/commits/fbdb17016a3674e234f6ecdbe672472ee405f52d

File: streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java]: I think we misunderstood each other. I was asking for an integration test that verifies that the rebalancing works for the maximum length of the tags. To see if the subscription has still a size that works. Maybe also in combination with many tasks.

Comment (lkokhreidze) [reply streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java]: Apologies, my mistake.
Makes sense. I've added `RackAwarenessIntegrationTest#shouldDoRebalancingWithMaximumNumberOfClientTags` that verifies rebalancing, as well as that rack awareness works after rebalancing too.

Comment (lkokhreidze) [reply streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java]: > Maybe also in combination with many tasks.

FYI - current topology generates 6 tasks, I guess that should be enough?

Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java]: I was more thinking about 1000 stateful tasks, because the task offset sums might also inflate the subscription.

Comment (lkokhreidze) [reply streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java]: Was able to push up to a two hundred stateful tasks and even that was already slow. After 500, test was getting flaky on my computer and sometimes it was just failing due to timeouts.

Feels like these kind of test is too heavy for a local integration tests. Any other way we could do it? Or any pointers to any of the existing integration tests that do the similar thing?

Anyway, I've pushed the changes. I also found one bug in the `ClientTagAwareStandbyTaskAssignor` which is now fixed.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (showuon) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: @cadonna , Good suggestion!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientTagAwareStandbyTaskAssignor.java
Comment (showuon) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientTagAwareStandbyTaskAssignor.java:63]: nice catch, and thanks for adding a test for it.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/StandbyTaskAssignmentUtilsTest.java
Comment (showuon) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/StandbyTaskAssignmentUtilsTest.java:99]: Thanks for the tests

File: streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java
Comment (showuon) [inline streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java]: 3 mins for a test is too long IMO. Could we reduce to 2 mins or less?

**Review by showuon — COMMENTED** [review]
@lkokhreidze , thanks for the update. Overall LGTM! Only 1 concern about the new added test, and its flakiness. Please help fix it. Thanks.

https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-11837/11/
```
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.streams.integration.RackAwarenessIntegrationTest.shouldDoRebalancingWithMaximumNumberOfClientTags
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.streams.integration.RackAwarenessIntegrationTest.shouldDoRebalancingWithMaximumNumberOfClientTags
    Build / JDK 8 and Scala 2.12 / org.apache.kafka.streams.integration.RackAwarenessIntegrationTest.shouldDoRebalancingWithMaximumNumberOfClientTags
    Build / JDK 8 and Scala 2.12 / org.apache.kafka.streams.integration.RackAwarenessIntegrationTest.shouldDoRebalancingWithMaximumNumberOfClientTags
    Build / JDK 17 and Scala 2.13 / org.apache.kafka.streams.integration.RackAwarenessIntegrationTest.shouldDoRebalancingWithMaximumNumberOfClientTags
    Build / JDK 17 and Scala 2.13 / org.apache.kafka.streams.integration.RackAwarenessIntegrationTest.shouldDoRebalancingWithMaximumNumberOfClientTags
```

**Comment by lkokhreidze** [comment]
Thanks @showuon for the feedback. Should be better now.
Also reduced test timeout to 1.5 minutes.

File: streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java]: I see. I asked some other people and apparently the size of the subscription is not that much of a concern currently. I would still propose to keep the test with 100 instead of 200 tasks. With 200 tasks, the test takes ~10 s on my laptop whereas with 100 tasks it takes ~6 s. I think 6 s is acceptable and might also reduce flakiness. @lkokhreidze and @showuon WDYT?

Comment (lkokhreidze) [reply streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java]: Sounds good to me. Will push changes with 100 stateful tasks for the rebalancing test.

**Review by cadonna — APPROVED** [review]
Thank you @lkokhreidze !

LGTM!

File: streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java
Comment (lkokhreidze) [reply streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java]: Pushed changes to reduce number of stateful tasks to 100
https://github.com/apache/kafka/pull/11837/commits/e6f717229acb4850aec1f33e13cd88cdc3ccb4bb

Comment (showuon) [reply streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java]: +1 for 100 tasks. Thanks.

Comment (lkokhreidze) [reply streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java]: Even with 100 tasks tests on CI failed with timeout. I've bumped timeout to 1.5 minutes in the latest commit.

**Comment by showuon** [comment]
@lkokhreidze , let's make sure the tests stable before we merge this final piece of PRs. Thanks.

File: streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java]: If after your PR the tests are still flaky, let's just test with 6 tasks and merge this PR. You can then add a system test for the scenario with larger subscriptions in a follow-up PR.

Comment (lkokhreidze) [reply streams/src/test/java/org/apache/kafka/streams/integration/RackAwarenessIntegrationTest.java]: Thanks @cadonna 
If no objections, I will do that now. The fact that it times out in 1 minute doesn't give much confidence that it won't be flaky.
I will follow up the system test.
fyi @showuon

**Comment by lkokhreidze** [comment]
@showuon 
I've reduced number of stateful tasks to 9. This should eliminate the flakiness. 
I will do the system test in the follow-up PR.

**Comment by cadonna** [comment]
@lkokhreidze We also need some updates to the docs. The sections to update are https://kafka.apache.org/31/documentation/streams/architecture#streams_architecture_recovery and maybe also https://kafka.apache.org/31/documentation/streams/developer-guide/config-streams.html. Not in this PR, please a new PR for the docs.

**Comment by lkokhreidze** [comment]
Thanks @cadonna ,
I assume it's okay to do it in a follow-up PR?

**Comment by cadonna** [comment]
Yes, doing it in a follow-up PR is totally fine!

**Review by showuon — APPROVED** [review]
LGTM! As long as the tests are stable, we can merge it. Thanks.

**Comment by lkokhreidze** [comment]
FYI, if I'm looking at test results correctly, seems like latest failures are not related.

```
Build / JDK 8 and Scala 2.12 / shouldRestoreState – org.apache.kafka.streams.integration.PurgeRepartitionTopicIntegrationTest
1m 3s
Build / JDK 17 and Scala 2.13 / [1] Type=Raft, Name=testElectionResultOutput, Security=PLAINTEXT – kafka.admin.LeaderElectionCommandTest
1m 10s
Build / JDK 17 and Scala 2.13 / shouldRestoreState – org.apache.kafka.streams.integration.PurgeRepartitionTopicIntegrationTest
1m 2s
Build / JDK 17 and Scala 2.13 / shouldRestoreState – org.apache.kafka.streams.integration.PurgeRepartitionTopicIntegrationTest
1m 3s
```

**Comment by cadonna** [comment]
Merged!
@lkokhreidze Congrats to this milestone! Nice work! 🥳

**Comment by lkokhreidze** [comment]
Big thank you @cadonna and @showuon for all the reviews.

**Comment by showuon** [comment]
Nice work @lkokhreidze ! And nice team work, @cadonna ! :)

**Pull request merged**
